homeassistant/components/motion_blinds/__init__.py

```diff
@@ -1,25 +1,71 @@
 """The motion_blinds component."""
-from asyncio import TimeoutError as AsyncioTimeoutError
+import asyncio
 from datetime import timedelta
 import logging
 from socket import timeout
 
+from motionblinds import MotionMulticast
+import voluptuous as vol
+
 from homeassistant import config_entries, core
-from homeassistant.const import CONF_API_KEY, CONF_HOST
+from homeassistant.const import (
+    ATTR_ENTITY_ID,
+    CONF_API_KEY,
+    CONF_HOST,
+    EVENT_HOMEASSISTANT_STOP,
+)
 from homeassistant.exceptions import ConfigEntryNotReady
-from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers import config_validation as cv, device_registry as dr
+from homeassistant.helpers.dispatcher import dispatcher_send
 from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
 
-from .const import DOMAIN, KEY_COORDINATOR, KEY_GATEWAY, MANUFACTURER
+from .const import (
+    ATTR_ABSOLUTE_POSITION,
+    ATTR_WIDTH,
+    DOMAIN,
+    KEY_COORDINATOR,
+    KEY_GATEWAY,
+    KEY_MULTICAST_LISTENER,
+    MANUFACTURER,
+    MOTION_PLATFORMS,
+    SERVICE_SET_ABSOLUTE_POSITION,
+)
 from .gateway import ConnectMotionGateway
 
 _LOGGER = logging.getLogger(__name__)
 
-MOTION_PLATFORMS = ["cover", "sensor"]
+CALL_SCHEMA = vol.Schema({vol.Required(ATTR_ENTITY_ID): cv.comp_entity_ids})
+
+SET_ABSOLUTE_POSITION_SCHEMA = CALL_SCHEMA.extend(
+    {
+        vol.Required(ATTR_ABSOLUTE_POSITION): vol.All(
+            cv.positive_int, vol.Range(max=100)
+        ),
+        vol.Optional(ATTR_WIDTH): vol.All(cv.positive_int, vol.Range(max=100)),
+    }
+)
+
+SERVICE_TO_METHOD = {
+    SERVICE_SET_ABSOLUTE_POSITION: {
+        "method": SERVICE_SET_ABSOLUTE_POSITION,
+        "schema": SET_ABSOLUTE_POSITION_SCHEMA,
+    }
+}
 
 
-async def async_setup(hass: core.HomeAssistant, config: dict):
+def setup(hass: core.HomeAssistant, config: dict):
     """Set up the Motion Blinds component."""
+
+    def service_handler(service):
+        method = SERVICE_TO_METHOD.get(service.service)
```

Comment:
service.service and method["method"] appear to be equivalent here.  Given that there is only one service at the moment, I might suggest unrolling this into a single hass.services.register call with a direct callback that implements handle_set_absolute_position then add the generality when adding a second service.

Reply:
I would rather keep it generic like this, this is also how it is done in other integration such as the denonavr (that also only has 1 service at the moment). In this way it will be easy to add aditional services later on.